### PR TITLE
Split unpacked 1D array

### PR DESCRIPTION
**Pull request by yTakatsukasa**
This is the initial step of [Issue 1656 ](https://www.veripool.org/issues/1656-Verilator--RFC-Relaxing-UNOPTFLAT-by-dividing-unpacked-array#change-5990).

Only unpacked 1D array can be splitted by this PR.

I will submit another PRs for further enhancement.

### CODE REVIEW

**Comment by wsnyder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/wsnyder/verilator/pullRequest?prid=4724658&bid=15773665)

File: src/V3SplitVar.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    // key:variable to be splitted. value:location where the variable is referenced.
+    vl_unordered_map<AstVar*, std::vector<AstArraySel*> > m_refs;
+    virtual void visit(AstNode* nodep) VL_OVERRIDE;
+    virtual void visit(AstNodeModule* nodep) VL_OVERRIDE;
+    virtual void visit(AstVar* nodep) VL_OVERRIDE;
+    virtual void visit(AstPragma* pragmap) VL_OVERRIDE;
+    virtual void visit(AstArraySel* nodep) VL_OVERRIDE;
+
+public:
+    SplitVarVisitor(AstNodeModule* modp);
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Codacy Issue found: [Class 'SplitVarVisitor' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/app/wsnyder/verilator/pullRequest?prid=4724658)

File: src/V3SplitVar.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        indexp = VN_CAST(constified, Const);
+    }
+
+    if (indexp) {  // OK
+        m_refs[varp].push_back(nodep);
+    } else {
+        nodep->bitp()->v3warn(SPLITVAR,
+                              "Variable \"" << vrefp->name() << "\" will not be splitted "
+                              "because index cannot be determined statically.");
+        if (m_refs.find(varp) != m_refs.end()) m_refs.erase(varp);
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Codacy Issue found: [Redundant checking of STL container element existence before removing it.](https://app.codacy.com/app/wsnyder/verilator/pullRequest?prid=4724658)

File: src/V3SplitVar.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        UINFO(3, "In module " << m_modp->name() << " var " << it->first->name() << " which has "
+                              << it->second.size() << " refs"
+                              << " will be splitted.\n");
+        AstVar* const varp = it->first;
+        AstUnpackArrayDType* const dtypep = VN_CAST(varp->subDTypep(), UnpackArrayDType);
+        std::vector<AstVar*> vars;
+        // Add the splitted variables
+        for (uint32_t i = 0; i < dtypep->arrayUnpackedElements(); ++i) {
+            std::stringstream name;
+            name << varp->name() << std::dec << i;
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: string name = varp->name + "__BRA__" + cvtToStr(i) + "__KET__"

Also for loop needs fix to index based on subscripts of user, change test to be non-zero and signed e.g. [5:-2].

Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: I have updated the test to have such indices.
V3Width:width() removes VAR_BASE attribute and make index 0-origin so the original behavior was also good. I added a comment about it.

**Comment by yTakatsukasa** [comment]
Thanks for the review. I have updated as you suggested in 8b38a15 .

**Comment by wsnyder** [comment]
Looking good, could you start in to add  in the other Issue 1656 stuff also to this pull?

**Comment by yTakatsukasa** [comment]
Packed data types and multi dimensional unpacked array are supported.

The remaining limitation is as written in the help and the source code.
  - unpacked array must be accessed via subscript. (e.g. `array[idx]`)

To relax the limitation, concatenate-like op to build an unpacked array is necessary.
I feel introducing the op to AST is not so simple because
it looks related to unsupported feature below and C++ code generation.
```
: Unsupported: Replication to form UNPACKARRAYDTYPE
              : ... In instance t
logic v [0:1] = { 1'b0, 1'b1};
                ^
```

So I'd like to separate the feature from this PR.

**Comment by wsnyder** [comment]
Looking good, I will review again shortly.

    : Unsupported: Replication to form UNPACKARRAYDTYPE
    logic v [0:1] = { 1'b0, 1'b1};

Instead declare the variable and unroll assignments e.g
     VAR v  (of original dtype)
      ASSIGN ARRAYSEL v 0,  new element 0 var
       ASSIGN ARRAYSEL v 0, new element 1

BTW please add a new test.pl file that uses the same Verilog (top_filename) but turns on --trace and does vcd_diff against a golden file to make sure it works, and that the brackets are properly shown in the trace var names.

**Comment by wsnyder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4724658&bid=15773665)

File: src/V3SplitVar.cpp [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+        if (m_firstRun) V3Stats::addStat("SplitVar, Split Unpacked Array", m_numSplit);
+    }
+    int numSplit() const { return m_numSplit; }
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // Check if the passed variable can be split.
+    // Even if this function returns true, the variable may not be split
+    // because the access to the variable cannot be determined statically.
+    static bool canSplit(const AstVar* nodep) {
+        if (AstNodeDType* dtypep = nodep->subDTypep()) {
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Codacy Issue found: [Variable 'dtypep' is assigned a value that is never used.](https://app.codacy.com/manual/wsnyder/verilator/pullRequest?prid=4724658)

**Comment by yTakatsukasa** [comment]
Thank you for reminding me of tracing.

### tracing
Now traced file is reasonable.
It is painful to see raw split fragmented variables, so concatenated result is dumped.
Test is added.

### assigning unpacked var
If the rhs is a simple VarRef, yes, multiple ASSIGN is enough.
When rhs is func/task call with a side effect (e.g. $display), unrolling does not help.
That's why I think concat-like op is necessary.

**Comment by yTakatsukasa** [comment]
fad2f6748b616fa2a91fda48159e6cb67feea9ff removes the remaining limitation (assignment of unpacked array).

File: src/V3SplitVar.cpp [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+        m_isLhs = false;
+        iterate(nodep->rhsp());
+    }
+    virtual void visit(AstAssignW* nodep) VL_OVERRIDE {
+        UASSERT_OBJ(m_isLhs == false, nodep, "unexpected nested assign");
+        m_isLhs = true;
+        iterate(nodep->lhsp());
+        m_isLhs = false;
+        iterate(nodep->rhsp());
+    }
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Seems suspicious you do Assign and AssignW but not NodeAssign, e.g. you don't do AssignDly's etc.

File: src/V3SplitVar.cpp [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+                                    consts[1]->toUInt()),
+                m_isLhs);
+        } else {
+            nodep->v3warn(SPLITVAR, "Variable "
+                                        << vrefp->prettyNameQ()
+                                        << " will not be split"
+                                           " because bit range cannot be determined statically.");
+            m_refs.erase(varp);
+        }
+    }
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: As noted should never happen, can assert

File: src/V3SplitVar.cpp [hunk 7]
```diff
@@ -0,0 +1,719 @@
...
+        : m_lastVarp(NULL) {}
+
+public:
+    static std::vector<std::pair<AstPragma*, AstVar*> > scan(AstNodeModule* modp) {
+        ScanPragmaVisitor v;
+        v.iterateChildren(modp);
+        return v.m_vars;
+    }
+};
+
```
[the remark this answers is not in the document]
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: In the initial implementation, it was a single pass.
But as the example around line 122, I found that 2-pass is necessary to catch the following case, VARREF appears before PRAGMA is found.
> logic [7:0] var0 = 8'b0; /* verilator split_var*/

When chekiing ASSIGN in INITIAL, VAR is not attached recognized as the target of split_var.

File: src/V3SplitVar.cpp (hunk 6 above)
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: When split_var is specified on a memory model, the accessed index is not constatnt. 
So this may happen and tested in t_split_var_1_bad.
> logic [1023:0] [7:0] memory;  /* verilator split_var */
> always @(posedge clk) begin
>    if (we) memory[waddr] <= wd;
>    rd <= memory[raddr];
> end

File: src/Verilator.cpp [hunk 8]
```diff
@@ -233,6 +239,10 @@ void process() {
     // Propagate constants into expressions
     V3Const::constifyAllLint(v3Global.rootp());
 
+    // Split packed variables into multiple pieces to resolve UNOPTFLAT.
+    // should be after constifyAllLint() which flattens to 1D bit vector
+    V3SplitVar::splitPackedVariable(v3Global.rootp());
+
```
[the remark this answers is not in the document]
Comment (yTakatsukasa) [reply src/Verilator.cpp]: Overall procedure have to be in the following order.
Unpacked arrays -> Packed variables -> further split fragments

I' afraid that moving splitPackedVariable() front increases complexity.
The splitting plan for packed variable internally compute accessed bit ranges and ranges are mapped on on a linear space.
It is possible to implement converting/interpreting multi-dimension to 1D in V3SplitVar. but may be somewhat duplicated implementation of V3Const.
Duplicated implementation among different .cpp will be more expensive than similar message text in the same .cpp in the future for maintenance.
So I think it is better to keep splitPackedVariable() after constifyAllLint().

I find less concern in moving splitUnpackedVariable() to just before splitPackedVariable() though some warnings that can be resolved by  splitPackedVariable() may appear.

File: src/V3SplitVar.cpp [hunk 9]
```diff
@@ -0,0 +1,719 @@
...
+    VL_DEBUG_FUNC;  // Declare debug()
+};
+
+//######################################################################
+// Split class functions
+void V3SplitVar::splitUnpackedVariable(AstNetlist* nodep) {
+    UINFO(2, __FUNCTION__ << ": " << endl);
+    // SplitUnpackedVarVisitor collapses one-dimension per one scan. so repeat until nothing to do.
+    for (int trial = 0, done = 0; done == 0; ++trial) {
+        {
```
[the remark this answers is not in the document]
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: Hmm, let me think about possibility to change so.

**Comment by yTakatsukasa** [comment]
I do have the same feeling, bigger task than I initially thought and appreciate your detailed review.

Other that I replied, I am updating per your suggestions in https://github.com/yTakatsukasa/verilator/tree/split_var0_dev. 
I will merge the update in the branch to this PR when I finish.

File: src/V3SplitVar.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: 74edc5cff4724f3f73246801a89243691aa71a48 does so.

718ebd4506a21e7fda79b88beed7b64cbc2d2eb9 also reduces AST traversal.

**Comment by yTakatsukasa** [comment]
Coding style stuff are fixed in e73f00ed9a92cca8daa296aaffebc0d71cd694f4 .
VL_DO_DANGLING is cherry picked from master.

File: test_regress/t/t_split_var_1_bad.pl [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+# redistribute it and/or modify it under the terms of either the GNU
+# Lesser General Public License Version 3 or the Perl Artistic License
+# Version 2.0.
+
+scenarios(simulator => 1);
+
+compile(
+    fails => 1,
+    verilator_flags2 => ['--stats '],
+    expect_filename => $Self->{golden_filename},
```
Comment (wsnyder) [inline test_regress/t/t_split_var_1_bad.pl]: Nit, extra space after stats.

File: bin/verilator [hunk 11]
```diff
@@ -4494,6 +4538,10 @@ With this change this assignment needs to be evaluated only once.  These
 sort of changes may also speed up your traditional event driven simulator,
 as it will result in fewer events per cycle.
 
+Another way to resolve this warning is to add split_var meta comment
+explained above. The variable will be split internnaly so that Verilator
+can optimize well.
```
Comment (wsnyder) [inline bin/verilator]: Another way to resolve this warning is to add the split_var meta comment
explained above. This will cause the variable to be split internally so that Verilator
can better optimize it.

File: bin/verilator [hunk 12]
```diff
@@ -4368,6 +4404,14 @@ Ignoring this warning may make Verilator simulations differ from other
 simulators, if the increased precision of real affects your model or DPI
 calls.
 
+=item SPLITVAR
+
+Warns that a variable with split_var metacomment was not split.
+Possible reasons are
+
+  The datatype is not supported for splitting. (e.g. real)
+  The access pattern of the variable can not be determined statically. (e.g. memory)
```
Comment (wsnyder) [inline bin/verilator]: Warns that a variable with the split_var metacomment was not split as
requested.  Possible reasons include:

* The datatype of the variable is not supported for splitting. (e.g. is a
real).

* The access pattern of the variable can not be determined
statically. (e.g. is accessed as a memory).

File: bin/verilator [hunk 13]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
 Same as C<sformat> in configuration files, see L</"CONFIGURATION FILES">
 for more information.
 
+=item /*verilator split_var*/
+
+Attached to a variable or a net declaration. It breaks the variable into
+multiple pieces. The following data types are supported.
+
+  Packed array and packed struct consist of Integer types (reg, logic, bit, byte, int...)
+  unpacked array of data type above
```
Comment (wsnyder) [inline bin/verilator]: Attached to a variable or a net declaration to break the variable into
multiple pieces to work around UNOPTFLAT and other optimization issues.
Typically the UNOPTFLAT warning will suggest which variables this might be
good to attach to.

Supports arrays and packed structs with integer types.

As adding split_var creates additional internal state, this may slow down
the Verilation and C compiler runtimes if used unnecessarily.

File: bin/verilator [hunk 14]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
...
+
+Verilator will internally convert a variable with the metacomment such as:
+
+    logic [7:0] x [0:1];  /*verilator split_var*/
+
+to packed variables below.
```
Comment (wsnyder) [inline bin/verilator]: to internal variables as if the code had the following:

File: bin/verilator [hunk 15]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
...
+
+    logic [7:0] x__BRA__0__KET__; /*verilator split_var*/
+    logic [7:0] x__BRA__1__KET__; /*verilator split_var*/
+
+Note that the generated packed variables still have split_var metacomment because
+they will be split into further smaller pieces accorting to each access pattern.
```
Comment (wsnyder) [inline bin/verilator]: Note that the generated packed variables retained the split_var metacomment
as they remain eligable to be split further into smaller pieces accorting
to the access patterns.

File: bin/verilator [hunk 16]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
...
+
+Packed variable will be split as the following example.
```
Comment (wsnyder) [inline bin/verilator]: Packed variables are similarly split, for example:

File: bin/verilator [hunk 17]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
...
+
+    logic [7:0] y;        /*verilator split_var*/
+
+to:
+
+    logic [1:0] y__BRA__7_6__KET__;
+    logic [5:0] y__BRA__5_0__KET__;
```
Comment (wsnyder) [inline bin/verilator]: ? shouldn't these also have the meta comment?

File: bin/verilator [hunk 18]
```diff
@@ -3405,6 +3405,42 @@ behavior.  See the test_regress/t/t_dpi_display.v file for an example.
...
+
+This operation helps optimization step to identify the dependencies
+among variables and may resolve UNOPTFLAT warning.
+If the warning is resolved, simulation speed gets improved.
+Please also see the explanation of UNOPTFLAT below.
+
+Splitting large array may slow donw the Verilating speed.
+
```
Comment (wsnyder) [inline bin/verilator]: Remove see above.

File: src/V3SplitVar.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply src/V3SplitVar.cpp]: The attr attachment should be able to be in AstAttrOf visitor in V3LinkParse.  This relies on the attributes being in a specific place where vars are declared, but there's no reason why split_var should have different rules than other attributes.

File: src/V3SplitVar.cpp [hunk 19]
```diff
@@ -0,0 +1,783 @@
...
+        }
+        return constp;
+    }
+    static int outerMostSizeOfUnpackedArray(AstVar* nodep) {
+        AstUnpackArrayDType* dtypep = VN_CAST(nodep->dtypep(), UnpackArrayDType);
+        UASSERT_OBJ(dtypep, nodep, "Must be unapcked array");
+        return dtypep->msb() - dtypep->lsb() + 1;
+    }
+
+    virtual void visit(AstNode* nodep) VL_OVERRIDE { iterateChildren(nodep); }
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: BTW if you'd like to make a separate pull request (off master) to add VL_OVERRIDE to all existing visitors that would be appreciated.

File: bin/verilator (hunk 17 above)
Comment (yTakatsukasa) [reply bin/verilator]: The metacomment is not necessary anymore because the bit slices are the final result and no visitor needs the comment.

Unpacked (with the metacomment) -> packed (with the metacomment) -> fragmented bit slice (without the metacomment)

**Comment by wsnyder** [comment]
Do you think this is ready to merge?  I updated this to the master, and will look through it all again.

**Comment by wsnyder** [comment]
On a large design I get: 
     ../V3SplitVar.cpp:602: rvalue is expected

The test case for this is huge, but believe this is only because the signal being split comes from a port in a lower module.

Can you please add a test for a split var at the "top", under a "source" and under a "dest" module?

Not sure what best way to deal with this is.... still would like to see everything done at same time (versus 2 places) if we can.  I think there's a lot of instance and assertion cases that may be problems and these should both be after V3Inst::dearrayAll.  Perhaps we move the ALWCOMBORDER warning to much later if that's the problem.

**Comment by yTakatsukasa** [comment]
Sorry for long stealth mode.
I have been testing your suggestion of running against all tests as in https://github.com/verilator/verilator/pull/2066#pullrequestreview-345328526 .
Now the test result is reasonable (failure is only on xml dump and some grep for statistics.)

I need to polish and add some more test cases, ETA is not so long.
During the testing,  splitting unpacked array got much robust and I hope to handle your case.

I am not so familiar with those layers and pre/post conditions.
After I finish clean the code, I will do some more experiment of the pass order.

**Comment by yTakatsukasa** [comment]
241e54e7b3cb4d14db3bd0ab667be4f8fa417737 splits all variables for testing.
77c52f8ab35c8e7a3db5454b7c8f256f00e5404f removes the code to mark all variables to be split.

I admit that splitting code for unpacked variable has a lot of changes, but it is robust and single-pass.

> Not sure what best way to deal with this is.... still would like to see everything done at same time (versus 2 places) if we can. 

I am wondering to which extent do you like to unify.
- single entry point in Verilator.cpp ?
- fuse SplitUnpackedVarVisitor and SplitPackedVarVisitor ?

I hesitate the latter one because each visit() function will have `if (isUnpackedVar(nodep)){` branches and will be difficult to track the codepath.

> I think there's a lot of instance and assertion cases that may be problems and these should both be after V3Inst::dearrayAll.

Do you mean you saw error cases related to arrayed instances ?

**Comment by wsnyder** [comment]
At a minimum single entry point into Verilator.  If you want to split unpacked, then packed (presumably in order is better) that's probably ok, but would prefer you make a list in the unpacked phase so you don't need to iterate everything a second time.  Each visitor pass is very time expensive on large designs.

>Do you mean you saw error cases related to arrayed instances ?

No, I just think it wouldn't work as it is now.  Actually based on what I mentioned above it doesn't work across some even non-arrayed instances, so that's a bigger issue, unless you just fixed it.

Found another problem, got a suggested split on a packed structure member (which is the correct suggestion).  We should allow this:

     typedef packed struct {
          logic a;  logic b;
     } foo_t;
     foo_t myvar; // verilator split_var          <<-- gets syntax error right now
     ...

**Comment by yTakatsukasa** [comment]
8e7366e uses attribute as suggested in https://github.com/verilator/verilator/pull/2066#discussion_r368571537 to achieve single pass for each visitor.
`split_var` metacomment needs to com before `;`.
The following example works properly as far as I checked dumped tree and debug message.

```
module t();
typedef struct packed {
    logic a;  logic b;
} foo_t;

foo_t myvar /* verilator split_var */;    

initial begin    
    myvar.a = 0;    
    myvar.b = 1;    
end
endmodule
```

While I tested with the custom version (241e54e ) which splits everything, `t_inst_array_partial.v` passed.
I edit `t_inst_array_partial.v` still passes even after I edit the file to add split_var metacomment to the module.


I will do some more experiment to see what will happen if  I change calling order in Verilator.cpp.

**Comment by yTakatsukasa** [comment]
If I change the order like this, 75 tests fail.

current: splitUnpackedVariable() -> undrivenAll() -> .... -> splitPackedVariable()
experiment: ... -> splitUnpackedVariable() -> splitPackedVariable() -> undrivenAll()

```
diff --git a/src/Verilator.cpp b/src/Verilator.cpp
index 63c47519..eedc8dc8 100644
--- a/src/Verilator.cpp
+++ b/src/Verilator.cpp
@@ -155,15 +155,6 @@ static void process() {
     // so V3Undriven sees variables to be eliminated, ie "if (0 && foo) ..."
     V3Const::constifyAllLive(v3Global.rootp());
 
-    // Split variables into multiple pieces to resolve UNOPTFLAT.
-    // This should come before undrivenAll() to eliminate ALWCOMBORDER warning
-    // which can be resolved by splitUnpackedVariable().
-    V3SplitVar::splitUnpackedVariable(v3Global.rootp());
-
-    // Signal based lint checks, no change to structures
-    // Must be before first constification pass drops dead code
-    V3Undriven::undrivenAll(v3Global.rootp());
-
     // Assertion insertion
     //    After we've added block coverage, but before other nasty transforms
     V3AssertPre::assertPreAll(v3Global.rootp());
@@ -180,10 +171,19 @@ static void process() {
     // Propagate constants into expressions
     V3Const::constifyAllLint(v3Global.rootp());
 
+    // Split variables into multiple pieces to resolve UNOPTFLAT.
+    // This should come before undrivenAll() to eliminate ALWCOMBORDER warning
+    // which can be resolved by splitUnpackedVariable().
+    V3SplitVar::splitUnpackedVariable(v3Global.rootp());
+
     // Split packed variables into multiple pieces to resolve UNOPTFLAT.
     // should be after constifyAllLint() which flattens to 1D bit vector
     V3SplitVar::splitPackedVariable(v3Global.rootp());
 
+    // Signal based lint checks, no change to structures
+    // Must be before first constification pass drops dead code
+    V3Undriven::undrivenAll(v3Global.rootp());
+
     if (!v3Global.opt.xmlOnly()) {
         // Remove cell arrays (must be between V3Width and scoping)
         V3Inst::dearrayAll(v3Global.rootp());
```

The following warning from V3Undriven.cpp caused most of failures.
> %Error-PROCASSWIRE: t/t_var_init.v:16: Procedural assignment to wire, perhaps intended var (IEEE 1800-2017 6.5): 'w2'

I will see the V3Undriven.cpp later, but I am afraid of dividing Undriven pass in the end.

I think Visitor cost after scopeAll() AST is expensive, but AST before scoping is much simple and have less nodes.

**Comment by yTakatsukasa** [comment]
I read V3Undriven.cpp. Now I think that the following change is enough.
- change the pass order
    - current : splitUnpackedVariable() -> undrivenAll() -> assertPreAll()..constifyAllLint() -> splitPackedVariable()
    - proposed: undrivenAll() -> assertPreAll()..constifyAllLint() -> splitVariable()
        - splitVariable() is merged function of splitUnpackedVariable() and splitPackedVariable()

Yes. Some split_var tests will get  `%Warning-ALWCOMBORDER:` warning which is false positive.
But the warning will be emitted regardless of V3SplitVar passes.

The checks in V3Undriven is a bit pessimistic, but seems inevitable to achieve the following comment in Verilator.cpp
- no changes to structures
- Must be before first constification pass drops dead code
 
So I will change V3Split to single traversal of AST as splitVariable() above.

**Comment by wsnyder** [comment]
Sounds like a great plan, thanks for keeping at it.

**Comment by yTakatsukasa** [comment]
a259f9d15dd5ba0374d9205dd6a3040606e3d036 has single entry point of V3SplitVar::splitVariable() and reuse some information from unpacked split.

**Comment by yTakatsukasa** [comment]
When splitting packed variables, e379683 fully reuse the information recorded in the preceding  process (splitting unpacked variables).

I think it's ready to be merged.

File: src/V3SplitVar.cpp [hunk 20]
```diff
@@ -0,0 +1,1258 @@
...
+        refs = visitor.getPackedVarRefs();
+    }
+    V3Global::dumpCheckGlobalTree("split_var", 0, v3Global.opt.dumpTreeLevel(__FILE__) >= 9);
+    { SplitPackedVarVisitor visitor(nodep, refs); }
+    V3Global::dumpCheckGlobalTree("split_var", 0, v3Global.opt.dumpTreeLevel(__FILE__) >= 9);
+}
+
+bool V3SplitVar::canSplitVar(const AstVar* varp) {
+    return !SplitUnpackedVarVisitor::cannotSplitReason(varp)
+           || !SplitPackedVarVisitor::cannotSplitReason(varp, false);
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Should the || be &&? If not please add comment to code as to why.

File: src/V3SplitVar.cpp [hunk 21]
```diff
@@ -0,0 +1,1258 @@
...
+#include "V3Stats.h"
+
+#include <algorithm>  // sort
+#include <set>
+#include <vector>
+#include VL_INCLUDE_UNORDERED_MAP
+#include VL_INCLUDE_UNORDERED_SET
+
+static AstNodeAssign* newAssign(FileLine* fileline, AstNode* lhs, AstNode* rhs,
+                                const AstVar* varp) {
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Here and below please add e.g. "split" prefix to all global names to disambiguate, as sometimes all .cpp files are compiled in a single compiler run and it keeps things cleaner.

I'd however prefer either:

A. move these into a class just used to wrap the implementation internals (SplitVarImp).
B. Or, make a base class and inherit it into each Visitor.

File: src/V3SplitVar.cpp [hunk 22]
```diff
@@ -0,0 +1,1258 @@
...
+    if (varp->isUsedLoopIdx()) return "it is used as a loop variable";
+    if (varp->isGenVar()) return "it is genvar";
+    if (varp->isParam()) return "it is parameter";
+    return NULL;
+}
+
+static const char* cannotSplitPackedVarReason(const AstVar* varp);
+
+template <class ALWAYSLIKE>
+void InsertBeginCore(ALWAYSLIKE* ap, AstNodeStmt* stmtp, AstNodeModule* modp) {
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: T_ALWAYSLIKE

File: src/V3SplitVar.cpp [hunk 23]
```diff
@@ -0,0 +1,1258 @@
...
+    if (ap->isJustOneBodyStmt() && ap->bodysp() == stmtp) {
+        stmtp->unlinkFrBack();
+        // Insert begin-end because temp value may be inserted to this block later.
+        const std::string name = "__SplitVar__Blk" + cvtToStr(modp->varNumGetInc());
+        ap->addStmtp(new AstBegin(ap->fileline(), name, stmtp));
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Always have a __V prefix when naming internal signals.

    "__VsplitVarBlk" + ...

File: src/V3SplitVar.cpp [hunk 24]
```diff
@@ -0,0 +1,1258 @@
...
+    }
+}
+
+void InsertBeginCore(AstInitial* initp, AstNodeStmt* stmtp, AstNodeModule* modp) {
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Function names shouldn't start with capitals.

File: src/V3SplitVar.cpp [hunk 25]
```diff
@@ -0,0 +1,1258 @@
...
+    } else if (AstAlwaysPublic* ap = VN_CAST(backp, AlwaysPublic)) {
+        InsertBeginCore(ap, stmtp, modp);
+    } else if (AstInitial* ap = VN_CAST(backp, Initial)) {
+        InsertBeginCore(ap, stmtp, modp);
+    }
+}
+
+// Assume startp is a port variable, traverse until the end of port declaration.
+// Returned pointer is the end of port variable.
+static AstNode* toEndOfPort(AstVar* startp) {
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Why can't you always insert right after the given port?

This adds a potential O(ports^2) iteration.

File: src/V3SplitVar.cpp [hunk 26]
```diff
@@ -0,0 +1,1258 @@
...
+    static int outerMostSizeOfUnpackedArray(AstVar* nodep) {
+        AstUnpackArrayDType* dtypep = VN_CAST(nodep->dtypep()->skipRefp(), UnpackArrayDType);
+        UASSERT_OBJ(dtypep, nodep, "Must be unapcked array");
+        return dtypep->msb() - dtypep->lsb() + 1;
+    }
+
+    void setContextAndIterateChildren(AstNode* nodep) {
+        const ContextKeeper keeper(&m_context, nodep);
+        iterateChildren(nodep);
+    }
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: I can live with this, but would probably favor the more verbose but obvious

        AstNode* origContextp = m_context;
        {
           m_contextp = nodep;
            iterateChildren(nodep);
        }
        m_contextp = origContextp;
    }

(Which is why I never made a template to save/restore such things.)

File: src/V3SplitVar.cpp [hunk 27]
```diff
@@ -0,0 +1,1258 @@
...
+//        if (some_cond) begin
+//            {packed_var__BRA__3__KET__, packed_var__BRA__2_0__KET__} = 4'b0;
+//        end else begin
+//            packed_var__BRA__3__KET__   = some_input0;
+//            packed_var__BRA__2_0__KET__ = some_input1;
+//        end
+//     end
+// </pre>
+//
+//
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Please add a bit more on what each pass does, and the data passed between them.

File: src/V3SplitVar.cpp [hunk 28]
```diff
@@ -0,0 +1,1258 @@
...
+    }
+
+public:
+    // If want to walk though the entrire AST.
+    explicit SplitPackedVarVisitor(AstNetlist* nodep)
+        : m_netp(nodep)
+        , m_modp(NULL)
+        , m_numSplit(0) {
+        iterate(nodep);
+    }
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Dead code?

File: src/V3SplitVar.cpp [hunk 29]
```diff
@@ -0,0 +1,1258 @@
...
+        , m_numSplit(0) {
+        for (RefsMap::iterator it = refs.begin(), it_end = refs.end(); it != it_end; ++it) {
+#if 1
+            m_modp = it->first;
+            it->second.visit(this);
+            split();
+            m_modp = NULL;
+#else
+            iterate(it->first);
+#endif
```
Comment (wsnyder) [inline src/V3SplitVar.cpp]: Dead code?

File: src/V3SplitVar.cpp (hunk 20 above)
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: `||` is correct.
If either SplitUnpackedVarVisitor or SplitPackedVarVisitor can handle it, that's fine.
I wrote in the comment.

File: src/V3SplitVar.cpp (hunk 25 above)
Comment (yTakatsukasa) [reply src/V3SplitVar.cpp]: Well, I don't remember well, but I guess AST looked clean and friendly to me.
Now the function is removed.

**Comment by yTakatsukasa** [comment]
Thanks for the comments.
a12d2f8 reflects your comments. (`git diff -w` shows smaller diff.)

**Comment by wsnyder** [comment]
Tried it on a few systems. Fixed some obvious stuff and pushed it here.  I get these on pre-C11 machines which still need cleanups:
```
../V3SplitVar.cpp:295: error: 'MapType::value_type' has not been declared
../V3SplitVar.cpp:295: error: ISO C++ forbids declaration of 'iterator' with no type
../V3SplitVar.cpp:295: error: typedef name may not be a nested-name-specifier
../V3SplitVar.cpp:295: error: expected ';' before 'SetIt'
../V3SplitVar.cpp: In member function 'void UnpackRefMap::registerVar(AstVar*)':
../V3SplitVar.cpp:312: error: 'MapType::value_type' has not been declared
../V3SplitVar.cpp: In member function 'void UnpackRefMap::swap(UnpackRefMap&)':
../V3SplitVar.cpp:334: error: 'class UnpackRefMap::MapType' has no member named 'swap'
cc1plus: warnings being treated as errors
../V3SplitVar.cpp: In member function 'size_t SplitUnpackedVarVisitor::collapse(UnpackRefMap&)':
../V3SplitVar.cpp:724: error: 'SetIt' is not a member of 'UnpackRefMap'
../V3SplitVar.cpp:724: error: expected ';' before 'sit'
../V3SplitVar.cpp:725: error: 'sit' was not declared in this scope
../V3SplitVar.cpp:725: error: 'sit_end' was not declared in this scope


```

**Comment by yTakatsukasa** [comment]
Thank you for the  touch ups and  letting me know the failure.
The error is fixed.

**Comment by wsnyder** [comment]
Pushed to git master towards 4.030.

Thanks for your great work on this contribution, as you know I've already put it to use!
